x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/DataFrameDataExtractor.java

```diff
@@ -93,6 +96,13 @@ public void cancel() {
         if (!hasNext()) {
             throw new NoSuchElementException();
         }
+
+        if (crossValidationSplitter.get() == null) {
+            // We lazily create the cross validation splitter as its creation
+            // may involve a search
+            crossValidationSplitter.set(context.crossValidationSplitterFactory.create());
+        }
```

Comment:
CachedSupplier is designed just for this type of scenario. Instead of a set once and a check, I would just use a `CachedSupplier<CrossValidationSplitter>`.